src/core/meshcop/tcat_agent.cpp

```diff
@@ -0,0 +1,519 @@
+                                     uint32_t     aOffset,
+                                     uint32_t     aLength,
+                                     bool        &aHasResponse)
+{
+    Error error;
+
+    OT_UNUSED_VARIABLE(aOutgoingMessage);
+    OT_UNUSED_VARIABLE(aHasResponse);
+
+    if (!IsCommandClassAuthorized(mCommissionerAuthorizationField.mCommissioningFlags,
+                                  mDeviceAuthorizationField.mCommissioningFlags))
+    {
+        error = kErrorRejected;
+        ExitNow();
+    }
+
+    switch (aTlvType)
+    {
+    case TlvType::kSetActiveOperationalDataset:
+        error = HandleActiveDataset(aIncommingMessage, aOffset, aLength);
+        break;
+
+    default:
+        error = kErrorInvalidCommand;
+    }
+
+exit:
+    return error;
+}
+
+Error TcatAgent::HandleExtraction(ot::Message &aIncommingMessage,
+                                  ot::Message &aOutgoingMessage,
+                                  uint8_t      aTlvType,
+                                  uint32_t     aOffset,
+                                  uint32_t     aLength,
+                                  bool        &aHasResponse)
+{
+    Error error;
+
+    OT_UNUSED_VARIABLE(aIncommingMessage);
+    OT_UNUSED_VARIABLE(aOutgoingMessage);
+    OT_UNUSED_VARIABLE(aOffset);
+    OT_UNUSED_VARIABLE(aLength);
+    OT_UNUSED_VARIABLE(aHasResponse);
+
+    switch (aTlvType)
+    {
+    default:
+        error = kErrorInvalidCommand;
+    }
+
+    // exit:
+    return error;
+}
+
+Error TcatAgent::HandleDecommissioning(ot::Message &aIncommingMessage,
+                                       ot::Message &aOutgoingMessage,
+                                       uint8_t      aTlvType,
+                                       uint32_t     aOffset,
+                                       uint32_t     aLength,
+                                       bool        &aHasResponse)
+{
+    Error error;
+
+    OT_UNUSED_VARIABLE(aIncommingMessage);
+    OT_UNUSED_VARIABLE(aOutgoingMessage);
+    OT_UNUSED_VARIABLE(aOffset);
+    OT_UNUSED_VARIABLE(aLength);
+    OT_UNUSED_VARIABLE(aHasResponse);
+
+    switch (aTlvType)
+    {
+    default:
+        error = kErrorInvalidCommand;
+    }
+
+    // exit:
+    return error;
+}
+
+Error TcatAgent::HandleApplication(ot::Message &aIncommingMessage,
+                                   ot::Message &aOutgoingMessage,
+                                   uint8_t      aTlvType,
+                                   uint32_t     aOffset,
+                                   uint32_t     aLength,
+                                   bool        &aHasResponse)
+{
+    Error error;
+
+    OT_UNUSED_VARIABLE(aOutgoingMessage);
+    OT_UNUSED_VARIABLE(aLength);
+    OT_UNUSED_VARIABLE(aHasResponse);
+
+    switch (aTlvType)
+    {
+    case kSendApplicationData:
+        LogInfo("Application data len:%d, offset:%d", aLength, aOffset);
+        aIncommingMessage.SetOffset(aOffset);
```

Comment:
I see why the `aIncomingMessage` is not `const` so we can change its `offset` here.

Suggest changing the `mAppDataReceiveCallback` so we pass the offset in the callback and avoid having to change it on `Message`?